lib/posix/clock.c

```diff
@@ -27,5 +47,74 @@ int clock_gettime(clockid_t clock_id, struct timespec *ts)
 	ts->tv_nsec = (s32_t) ((elapsed_msecs % MSEC_PER_SEC) *
 					USEC_PER_MSEC * NSEC_PER_USEC);
 
+	ts->tv_sec += base.tv_sec;
+	ts->tv_nsec += base.tv_nsec;
+	if (ts->tv_nsec > NSEC_PER_SEC) {
+		ts->tv_sec++;
+		ts->tv_nsec -= NSEC_PER_SEC;
+	}
+
 	return 0;
 }
+
+/**
+ * @brief Set the time of the specified clock.
+ *
+ * See IEEE 1003.1.
+ *
+ * Note that only the `CLOCK_REALTIME` clock can be set using this
+ * call.
+ */
+int clock_settime(clockid_t clock_id, const struct timespec *tp)
+{
+	struct timespec base;
+	int res;
+
+	if (clock_id != CLOCK_REALTIME) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	res = clock_gettime(clock_id, &base);
+	if (res != 0) {
+		return res;
+	}
+
+	base.tv_sec = tp->tv_sec - base.tv_sec;
+	base.tv_nsec = tp->tv_nsec - base.tv_nsec;
+
+	/* Make the tv_nsec field positive.  Only try a little,
+	 * assuming that for both clocks, the value is in a reasonable
+	 * range. */
```

Comment:
Normally, I'm not one for over-checking user arguments.  But in this case a broken argument to clock_settime() is going to leave the clock API in a corrupt state, returning crazy answers forever.  The 64 bit code to compute the subtraction (in units of nanoseconds) is simpler/cleaner than hand-modulus anyway, not that much larger or slower at runtime, and it's not like clock_settime() is ever going to be a performance path.  Can we do this right?